Test/DafnyTests/TestAttribute.dfy

```diff
@@ -67,3 +67,205 @@ method {:test} PassingTestUsingNoLHSAssignOrHalt() {
+
+method {:synthesize} MockValidEven() returns (e:Even) 
+    ensures fresh(e) 
+    ensures e.IsValid() == true
+method {:synthesize} MockInValidEven() returns (e:Even) 
+    ensures fresh(e) 
+    ensures e.IsValid() == false
+
+method {:test} PassingTestUsingValidMock() {
+    var e:Even := MockValidEven();
+    expect(e.IsValid());
+}
+
+method {:test} PassingTestUsingInValidMock() {
+    var e:Even := MockInValidEven();
+    expect(!e.IsValid());
+}
+
+method {:synthesize} MockSum() returns (e:Even) 
+    ensures fresh(e) 
+    ensures e.Sum(2, 2) == 3
+
+method {:test} PassingTestMockSum() {
+    var e:Even := MockSum();
+    expect(e.Sum(2, 2) == 3);
+}
+
+method {:synthesize} MockSumForall() returns (e:Even) 
+    ensures fresh(e) 
+    ensures forall a:int, b:int :: e.Sum(a, b) == 3
+
+method {:test} PassingTestMockForall() {
+    var e:Even := MockSumForall();
+    expect(e.Sum(2, 2) == 3);
+    expect(e.Sum(3, 2) == 3);
+}
+
+method {:synthesize} MockSumAsMultiplication() returns (e:Even) 
+    ensures fresh(e) 
+    ensures forall a:int :: e.Sum(3, a) == a * 3
+    
+method {:test} PassingTestMockSumAsMultiplication() {
+    var e:Even := MockSumAsMultiplication();
+    expect(e.Sum(3, 2) == 6);
+    expect(e.Sum(3, 0) == 0);
+}
+
+method {:synthesize} MockSumWithArgumentMatcher() returns (e:Even) 
+    ensures fresh(e) 
+    ensures forall a:int, b:int :: (b < a) ==> (e.Sum(a, b) == a * b)
+    ensures forall a:int, b:int :: (b >= a) ==> (e.Sum(a, b) == -a * b)
+    
+method {:test} PassingTestMockSumWithArgumentMatcher() {
+    var e:Even := MockSumWithArgumentMatcher();
+    expect(e.Sum(2, 2) == -4);
+    expect(e.Sum(2, 3) == -6);
+    expect(e.Sum(4, 0) == 0);
+    expect(e.Sum(5, 1) == 5);
+}
+
+method {:synthesize} MockField() returns (e:Even) 
+    ensures fresh(e) 
+    ensures e.value == 7;
+    
+method {:test} PassingTestMockField() {
+    var e:Even := MockField();
+    e.value := 5;
+    expect(e.value == 7);
+    expect(e.value != 5);
+}
+
+method {:synthesize} ParametrizedMock(a: int) returns (e:Even) 
+    ensures fresh(e) 
+    ensures e.value == a;
+    
+method {:test} PassingTestParameterizedMock() {
+    var e:Even := ParametrizedMock(24);
+    expect(e.value == 24);
+    expect(e.value != 7);
+}
+
+method {:synthesize} SelfReferentialMock() returns (e:Even) 
+    ensures fresh(e)
+    ensures e.Next() == e
+    
+method {:test} PassingTestSelfReferentialMock() {
+    var e:Even:= SelfReferentialMock();
+    expect(e.Next() == e);
+    expect(e.Next() != null);
+}
+
+method {:synthesize} CrossReferentialMock() returns (e1:Even, e2:Even) 
+    ensures fresh(e1) && fresh(e2) 
+    ensures e1.Next() == e2
+    ensures e2.Next() == e1
+    
+method {:test} PassingTestCrossReferentialMock() {
+    var e1:Even, e2:Even := CrossReferentialMock();
+    expect(e1.Next() == e2);
+    expect(e1.Next() != e1);
+    expect(e2.Next() == e1);
+    expect(e2.Next() != e2);
+}
+
+class StringMap {
+
+	var m:map<string, string>;
```

Comment:
nit: the indentation on this file is all over the place. :) We've generally settled on a two-space indent.

Reply:
Have changed everything to two-space indent.